### ISSUE TITLE
Add hs.websocket

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#2295]
Currently there's a number of issues related to websockets - specifically:

- [Add documentation for `hs.http.websocket`](https://github.com/Hammerspoon/hammerspoon/issues/2228)
- [`hs.http.websocket` crashes hammerspoon if send is called too soon](https://github.com/Hammerspoon/hammerspoon/issues/2230)
- [`hs.http.websocket` needs onopen and onclose callbacks](https://github.com/Hammerspoon/hammerspoon/issues/2229)
- [How to use websocket schemes `ws://` or `wss://`](https://github.com/Hammerspoon/hammerspoon/issues/2056)

Given that `hs.http.websocket` is currently undocumented (at least until #2293 lands), this is probably a good time to review the websocket functionality and see if it can be improved.

As discussed in #2292, my attempts to use `hs.http.websocket` have been fairly unsuccessful, as I keep running into this error:

```lua
2020-01-17 02:00:48: 02:00:48    LuaSkin: (secondary thread): hs.http.webSocket: didFail: Error Domain=PSWebSocketErrorDomain Code=3 "Output stream end encountered" UserInfo={NSLocalizedDescription=Output stream end encountered}
```

...which seems to be a bug/issue in the [PocketSocket](https://github.com/zwopple/PocketSocket/issues/57) Framework.

[PocketSocket](https://github.com/zwopple/PocketSocket) hasn't been updated since 14 Aug 2017, so I wonder if it's worthwhile creating a new `hs.websocket` extension that replaces `hs.http.websocket`, and potentially uses a more recently updated websocket framework?

[SocketRocket](https://github.com/facebook/SocketRocket) seems like a popular alternative, but it also hasn't been updated since 29 Apr 2018, and currently has a whole heap of pull requests awaiting review.

[Jetfire](https://github.com/acmacalister/jetfire) also seems popular, but again hasn't been updated since 12 Aug 2017. However, interestingly sonos [forked it](https://github.com/sonos/jetfire) on 20 Oct 2018.

The [Pusher Channels websocket client](https://github.com/pusher/pusher-websocket-swift) looks like it's actively developed and also works with Objective-C and macOS - so this is something we could potentially fork and modify? This probably falls into the too-hard basket though.

Another idea I had was to do the websocket work in NodeJS and use something like [NodObjC](https://github.com/TooTallNate/NodObjC) as a bridge, but again that probably complicates things a bit too much.

Long story short... my main focus for now is trying to "convert" this [Loupedeck CT Node.js module](https://github.com/bitfocus/loupedeck-ct), into something I can use in Hammerspoon, and currently my attempts using `hs.http.websocket` have failed (whilst it all works great in NodeJS-land), so I'd love to come up with a solution.

Based on some existing open issues, it seems websockets are important to quite a few users in the Hammerspoon community, so very much open to ideas and suggestions.

Thoughts @asmagill @jeffmikels @Robert-M-Muench @seanfarley ?

**Comment by latenitefilms** [comment]
FYI: I've been playing around with SocketRocket, and thus far it seems to work as intended

**Comment by latenitefilms** [comment]
@asmagill - There's one thing I can't work out yet, which I'm sure you'll know the answer to.

In NodeJS land, I can send this message to the Socket and it works as intended:

```js
let buf = Buffer.from("\x03\x03\x06"); // get serial
this.connection.send(buf);
```

...however, I'm not sure how to do this correctly in Lua-land.

I thought that `ws:send("\3\3\6")` should do the job, but it's not giving me the response I need.

I also tried:

```lua
function utf8_from(t)
  local bytearr = {}
  for _, v in ipairs(t) do
    local utf8byte = v < 0 and (0xff + v + 1) or v
    table.insert(bytearr, string.char(utf8byte))
  end
  return table.concat(bytearr)
end
local result = utf8_from({3,3,6})
print(hs.utf8.hexDump(result))
ws:send(result)
```

I assume something is going funky going from Lua string to NSString to NSData?

**Comment by asmagill** [comment]
I know you've talked about changing websocket into its own module and/or changing the backend pod which handles the actual websocket work for you, so line number/file may be different, but in the lua/objc bridge portion (i.e. the function `http_ws_send` of the core `hs.http` internal.m file) line 685 reads: `NSString *message = [skin toNSObjectAtIndex:2];`

This usage of `toNSObjectAtIndex:` with no options specified will *always* force a lua string to be strictly valid and printable (i.e. proper UTF8), even if it means converting bad sequences into the UTF "unprintable" symbol (i.e. it doesn't promise an *exact* copy of the input).

If you want to make sure that it's treated as data and no potential internal munging occurs, you should use `NSData *message = [skin toNSObjectAtIndex:2 withOptions:LS_NSLuaStringAsDataOnly];`

Of course, this assumes that the websocket pod can accept an NSData object as the message (or at least a `const void*` which can be achieved with `message.bytes`, assuming that `message` is an NSData object); if not, then we need a different pod because the pod itself has inherent problems with raw data.

You'll also need to scan the appropriate module to see if there are other methods that *should* accept raw data that need to be adjusted as well.

In short:

`[skin toNsObjectAtIndex:x]` - always returns an NSString, but may have changed the data to make it a proper string. Terminates at `\0` in the lua string, even if what follows would be considered valid.

`[skin toNSObjectAtIndex:x withOptions:LS_NSPreserveLuaStringExactly]` - will return an NSString object if (and *only* if) the full lua string contains proper and valid UTF8 sequences; otherwise returns NSData. An embedded `\0` anywhere but the end would cause the returned value to be NSData.

`[skin toNSObjectAtIndex:x withOptions: LS_NSLuaStringAsDataOnly]` - always returns an NSData, but guarantees that no munging or changing occurs. Handles `'\0' as just another byte in the sequence.

I generally find its easiest to use the last version described above if I want to accept raw data (though there are exceptions, if the final use of the result has separate methods optimized for NSString and NSData); if you use the second version, you have to check the return type to determine if it's actually an NSString or an NSData object and then cast is appropriately (e.g. `if ([message isKindOfObject:[NSData class]]) ... // make sure to store message in something of type NSData or use as (NSData *)message`).

**Comment by asmagill** [comment]
Though looking closer, `\3\3\6` *should* be valid, though "invisible" UTF8, so not sure if that would fix the specific issue; however, making the change suggested above is going to be necessary anyways for proper handling of raw data, so let me know if it helps or not; I'll try to update my Hammerspoon later tonight and see what branch you're currently working with so I can try and test things directly myself as well)

**Comment by latenitefilms** [comment]
@asmagill - Thank you, thank you, thank you! Genius as usual! Changing to `[skin toNSObjectAtIndex:x withOptions: LS_NSLuaStringAsDataOnly]` has fixed my issue. Woohoo!
